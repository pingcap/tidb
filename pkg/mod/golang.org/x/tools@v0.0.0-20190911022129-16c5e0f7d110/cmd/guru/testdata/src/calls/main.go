package main

import (
	"fmt"
)

// Tests of call-graph queries.
// See go.tools/guru/guru_test.go for explanation.
// See calls.golden for expected query results.

func A(x *int) { // @pointsto pointsto-A-x "x"
	// @callers callers-A "^"
	// @callstack callstack-A "^"
}

func B(x *int) { // @pointsto pointsto-B-x "x"
	// @callers callers-B "^"
}

func foo() {
}

// apply is not (yet) treated context-sensitively.
func apply(f func(x *int), x *int) {
	f(x) // @callees callees-apply "f"
	// @callers callers-apply "^"
}

// store *is* treated context-sensitively,
// so the points-to sets for pc, pd are precise.
func store(ptr **int, value *int) {
	*ptr = value
	// @callers callers-store "^"
}

func call(f func() *int) {
	// Result points to anon function.
	f() // @pointsto pointsto-result-f "f"

	// Target of call is anon function.
	f() // @callees callees-main.call-f "f"

	// @callers callers-main.call "^"
}

func main() {
	var a, b int
	go apply(A, &a) // @callees callees-main-apply1 "app"
	defer apply(B, &b)

	var c, d int
	var pc, pd *int // @pointsto pointsto-pc "pc"
	store(&pc, &c)
	store(&pd, &d)
	_ = pd // @pointsto pointsto-pd "pd"

	call(func() *int {
		// We are called twice from main.call
		// @callers callers-main.anon "^"
		return &a
	})

	// Errors
	_ = "no function call here"   // @callees callees-err-no-call "no"
	print("builtin")              // @callees callees-err-builtin "builtin"
	_ = string("type conversion") // @callees callees-err-conversion "str"
	call(nil)                     // @callees callees-err-bad-selection "call\\(nil"
	if false {
		main() // @callees callees-err-deadcode1 "main"
	}
	var nilFunc func()
	nilFunc() // @callees callees-err-nil-func "nilFunc"
	var i interface {
		f()
	}
	i.f() // @callees callees-err-nil-interface "i.f"

	i = new(myint)
	i.f() // @callees callees-not-a-wrapper "f"

	// statically dispatched calls. Handled specially by callees, so test that they work.
	foo()         // @callees callees-static-call "foo"
	fmt.Println() // @callees callees-qualified-call "Println"
	m := new(method)
	m.f() // @callees callees-static-method-call "f"
	g := new(embeddedIface)
	g.iface = m
	g.f() // @callees callees-implicit-selection-method-call "f"
}

type myint int

func (myint) f() {
	// @callers callers-not-a-wrapper "^"
}

type method int

func (method) f() {
}

type embeddedIface struct {
	iface
}

type iface interface {
	f()
}

var dynamic = func() {}

func deadcode() {
	main() // @callees callees-err-deadcode2 "main"
	// @callers callers-err-deadcode "^"
	// @callstack callstack-err-deadcode "^"

	// Within dead code, dynamic calls have no callees.
	dynamic() // @callees callees-err-deadcode3 "dynamic"
}

// This code belongs to init.
var global = 123 // @callers callers-global "global"

// The package initializer may be called by other packages' inits, or
// in this case, the root of the callgraph.  The source-level init functions
// are in turn called by it.
func init() {
	// @callstack callstack-init "^"
}
