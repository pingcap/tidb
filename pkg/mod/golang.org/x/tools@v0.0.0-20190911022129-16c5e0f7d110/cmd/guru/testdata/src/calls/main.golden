-------- @pointsto pointsto-A-x --------
this *int may point to these objects:
	a
	b

-------- @callstack callstack-A --------
Found a call path from root to calls.A
calls.A
dynamic function call from calls.apply
concurrent static function call from calls.main

-------- @pointsto pointsto-B-x --------
this *int may point to these objects:
	a
	b

-------- @callers callers-B --------
calls.B is called from these 1 sites:
	dynamic function call from calls.apply

-------- @callees callees-apply --------
this dynamic function call dispatches to:
	calls.A
	calls.B

-------- @callers callers-apply --------
calls.apply is called from these 2 sites:
	concurrent static function call from calls.main
	deferred static function call from calls.main

-------- @callers callers-store --------
calls.store is called from these 2 sites:
	static function call from calls.main
	static function call from calls.main

-------- @pointsto pointsto-result-f --------
this func() *int may point to these objects:
	calls.main$1

-------- @callees callees-main.call-f --------
this dynamic function call dispatches to:
	calls.main$1

-------- @callers callers-main.call --------
calls.call is called from these 2 sites:
	static function call from calls.main
	static function call from calls.main

-------- @callees callees-main-apply1 --------
this static function call dispatches to:
	calls.apply

-------- @pointsto pointsto-pc --------
this *int may point to these objects:
	c

-------- @pointsto pointsto-pd --------
this *int may point to these objects:
	d

-------- @callees callees-err-no-call --------

Error: there is no function call here
-------- @callees callees-err-builtin --------

Error: this is a call to the built-in 'print' operator
-------- @callees callees-err-conversion --------

Error: this is a type conversion, not a function call
-------- @callees callees-err-bad-selection --------

Error: ambiguous selection within function call (or conversion)
-------- @callees callees-err-deadcode1 --------
this static function call dispatches to:
	calls.main

-------- @callees callees-err-nil-func --------
dynamic function call on nil value

-------- @callees callees-err-nil-interface --------
dynamic method call on nil value

-------- @callees callees-not-a-wrapper --------
this dynamic method call dispatches to:
	(calls.myint).f

-------- @callees callees-static-call --------
this static function call dispatches to:
	calls.foo

-------- @callees callees-qualified-call --------
this static function call dispatches to:
	fmt.Println

-------- @callees callees-static-method-call --------
this static function call dispatches to:
	(calls.method).f

-------- @callees callees-implicit-selection-method-call --------
this dynamic method call dispatches to:
	(calls.method).f

-------- @callers callers-not-a-wrapper --------
(calls.myint).f is called from these 1 sites:
	dynamic method call from calls.main

-------- @callees callees-err-deadcode2 --------
this static function call dispatches to:
	calls.main

-------- @callstack callstack-err-deadcode --------
calls.deadcode is unreachable in this analysis scope

-------- @callees callees-err-deadcode3 --------

Error: this call site is unreachable in this analysis
-------- @callers callers-global --------
calls.init is called from these 1 sites:
the root of the call graph

-------- @callstack callstack-init --------
Found a call path from root to calls.init#1
calls.init#1
static function call from calls.init

